genrl/deep/agents/dqn/values.py

```diff
@@ -0,0 +1,237 @@
+        state_dim: int,
+        action_dim: int,
+        fc_layers: Tuple = (128,),
+        noisy_layers: Tuple = (128, 512),
+        num_atoms: int = 51,
+    ):
+        super(MlpCategoricalValue, self).__init__()
+
+        self.action_dim = action_dim
+        self.num_atoms = num_atoms
+
+        self.model = noisy_mlp(
+            [state_dim] + list(fc_layers),
+            list(noisy_layers) + [self.action_dim * self.num_atoms],
+        )
+
+    def forward(self, state: torch.Tensor) -> torch.Tensor:
+        features = self.model(state)
+        dist = nn.functional.softmax(features.view(-1, self.num_atoms), dim=0).view(
+            -1, self.action_dim, self.num_atoms
+        )
+        return dist
+
+    def reset_noise(self) -> None:
+        for layer in self.model:
+            if isinstance(layer, NoisyLinear):
+                layer.reset_noise()
+
+
+class CnnCategoricalValue(nn.Module):
+    """Class for Noisy DQN's MLP Q-Value function
+
+    Attributes:
+        framestack (int): No. of frames being passed into the Q-value function
+        action_dim (int): Action space dimensions
+        fc_layers (:obj:`tuple`): Fully connected layer dimensions
+        noisy_layers (:obj:`tuple`): Noisy layer dimensions
+        num_atoms (int): Number of atoms used to discretise the
+            Categorical DQN value distribution
+    """
+
+    def __init__(
+        self,
+        framestack: int,
+        action_dim: int,
+        fc_layers: Tuple = (128, 128),
+        noisy_layers: Tuple = (128, 512),
+        num_atoms: int = 51,
+    ):
+        super(CnnCategoricalValue, self).__init__()
+
+        self.action_dim = action_dim
+        self.num_atoms = num_atoms
+
+        self.conv, output_size = cnn((framestack, 16, 32))
+        self.model = noisy_mlp(
+            [output_size] + list(fc_layers),
+            list(noisy_layers) + [self.action_dim * self.num_atoms],
+        )
+
+    def forward(self, inp: torch.Tensor) -> torch.Tensor:
+        inp = self.conv(inp)
+        inp = inp.view(inp.size(0), -1)
+        inp = self.model(inp)
+        inp = nn.functional.softmax(inp.view(-1, self.num_atoms)).view(
+            -1, self.action_dim, self.num_atoms
+        )
+        return inp
+
+    def reset_noise(self) -> None:
+        for layer in self.model:
+            if isinstance(layer, NoisyLinear):
+                layer.reset_noise()
+
+
+dqn_value_registry = {
+    "mlpdueling": MlpDuelingValue,
+    "cnndueling": CnnDuelingValue,
+    "mlpnoisy": MlpNoisyValue,
+    "cnnnoisy": CnnNoisyValue,
+    "mlpcategorical": MlpCategoricalValue,
+    "cnncategorical": CnnCategoricalValue,
+}
+
+
+def get_dqn_value_from_name(name_: str):
+    """
+    Returns DQN model given the type of DQN
+
+    :param name_: Name of the policy needed
+    :type name_: str
+    :returns: DQN class to be used
+    """
+    if name_ in dqn_value_registry:
+        return dqn_value_registry[name_]
+    raise NotImplementedError
```

Comment:
I definitely like this, but I'd prefer these to be part of general value function definitions since at the end of the day, they can be used right?

Reply:
It's just that there are 6 different functions. The values.py file would become really big. Also, can we use them anywhere else? (Except in DQN) 

Also, I was also thinking that this seemed redundant to develop so many different classes and was thinking of simplifying this too into snippets. (Otherwise how do you use MlpDuelingValue and MlpNoisyValue together?) It would be really messy if we had to implement Rainbow and had to make a MlpDuelingNoisyCategoricalValue :p

Reply:
I'm not sure about Dueling but noisy networks are pretty common to enhance exploration. For e.g. Categorical can be used in Distributional DDPG?

Reply:
Ok, I'll move these if `values.py` doesn't get too big, otherwise I'll see what else we can do.